content_sync/tasks.py

```diff
@@ -392,3 +394,119 @@ def check_incomplete_publish_build_statuses():
             log.exception(
                 "Error updating publishing status for website %s", website.name
             )
+
+
+@app.task(acks_late=True)
+def update_websites_in_root_website():
+    """
+    Get all websites published to draft / live at least once, and for each one create or update
+    a WebsiteContent object of type website in the website denoted by settings.ROOT_WEBSITE_NAME
+    """
+    if settings.CONTENT_SYNC_BACKEND:
+        root_website = Website.objects.get(name=settings.ROOT_WEBSITE_NAME)
+        # Get all sites, minus any sites that have never been successfully published
+        sites = Website.objects.exclude(
+            Q(**{"draft_publish_date__isnull": True})
+            & Q(**{"publish_date__isnull": True})
+        )
+        sites = sites.exclude(Q(url_path__isnull=True))
+        # Exclude the root website
+        sites = sites.exclude(name=settings.ROOT_WEBSITE_NAME)
+        fields = [
+            "website",
+            "type",
+            "title",
+            "dirpath",
+            "filename",
+            "is_page_content",
+            "metadata",
+        ]
+        with WebsiteContent.bulk_objects.bulk_update_or_create_context(
+            fields, match_field="filename", batch_size=100
+        ) as bulk_update:
+            for website in sites:
+                site_metadata = WebsiteContent.objects.get(
+                    website=website, type="sitemetadata"
+                ).metadata
+                # We want this content to show up in lists, but not render pages
+                site_metadata["_build"] = {"list": True, "render": False}
+                # Set the content to draft if the site has not been published or is unpublished
+                if website.unpublish_status is not None or website.publish_date is None:
+                    site_metadata["draft"] = True
+                # Carry over url_path for proper linking
+                site_metadata["url_path"] = website.url_path
+                bulk_update.queue(
+                    WebsiteContent(
+                        website=root_website,
+                        type="website",
+                        title=website.title,
+                        dirpath=WEBSITE_LISTING_DIRPATH,
+                        filename=website.short_id,
+                        is_page_content=True,
+                        metadata=site_metadata,
+                    )
+                )
+        website_content = WebsiteContent.objects.filter(
+            website=root_website, type="website"
+        )
+        backend = api.get_sync_backend(website=root_website)
+        backend.sync_all_content_to_backend(query_set=website_content)
+
+
+@app.task(acks_late=True)
+def update_website_in_root_website(website, version):
```

Comment:
I tried various ways of accomplishing this all day today and hit a brick wall of circular imports at every turn. `content_sync.api` calls a number of functions from `websites.api` already, so importing anything from `content_sync` causes a circular import error. I tried creating a new `tasks.py` file in `websites`, adding the `update_website_in_root_website` and `remove_website_in_root_website` as simple functions rather than tasks inside `websites/api.py`. The only way I could get around the error was to modify `update_website_in_root_website` to get the sync backend and the `publish_website` function manually with `import_string`, but then the function became untestable because at that point you can't mock getting the sync backend. Overall, I don't think it's worth the trouble to try and call it during `update_website_status` as even if the pipeline fails and the `WebsiteContent` object is created, the pipeline will likely be fixed eventually and the site published. If the site is unpublished, the `WebsiteContent` object will be removed anyway. If you can think of something stupid I might be doing wrong, or a better way to accomplish this while avoiding the circular import please let me know.

Reply:
I'm fine with leaving it as is then, shouldn't happen often and as you say the pipeline should successfully complete eventually.